File: UI/Validation/DateTimeValidationRule.fs
namespace UI.Validation

open System
open System.Globalization
open System.Windows.Controls

type DateTimeValidationRule() =
    inherit ValidationRule()
        
        override x.Validate(value:obj, cultureInfo:CultureInfo) =
            let text = value :?> string
            let pattern = "yyyy-MM-dd HH:mm:ss"
            let culture = CultureInfo.CurrentCulture
            let style = DateTimeStyles.None
            match DateTime.TryParseExact(text, pattern, culture, style) with
            | true, x -> new ValidationResult(true, null)
            | _ -> new ValidationResult(false, "Date Expected: e.g. 2016-01-18T13:45:00.000")

File: UI.CSharp/Validation/DateTimeValidationRule.cs
using System;
using System.Globalization;
using System.Windows.Controls;

namespace UI.Validation {
    public class DateTimeValidationRule : ValidationRule {

        public override ValidationResult Validate(object value, CultureInfo cultureInfo) {
            DateTime output;
            if (DateTime.TryParseExact(value as string, "yyyy-MM-dd HH:mm:ss", CultureInfo.CurrentCulture, DateTimeStyles.None, out output)) {
                return new ValidationResult(true, null);
            } else {
                return new ValidationResult(false, "Date Expected: e.g. 2016-01-18T13:45:00.000");
            }
        }
    }
}
